File: bookService/src/main/java/org/library/bookservice/model/Book.java
package org.library.bookservice.model;

import jakarta.persistence.*;
import lombok.Data;
import org.library.bookservice.model.base.Archivable;
import org.library.bookservice.model.base.Identifiable;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
public class Book implements Identifiable, Archivable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;
    private String title;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "author_id")
    private Author author;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @Column(length = 500)
    private String description;

    @Column(unique = true, length = 20)
    private String ISBN;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "publisher_id")
    private Publisher publisher;

    @Column(name = "release_date")
    private LocalDate releaseDate;

    @Column(nullable = false)
    private Double price;

    @ManyToMany
    @JoinTable(
            name = "book_genre",
            joinColumns = @JoinColumn(name = "book_id"),
            inverseJoinColumns = @JoinColumn(name = "genre_id")
    )
    private List<Genre> bookGenres = new ArrayList<>();

    @Column(nullable = false, columnDefinition = "TINYINT(1)")
    private boolean archived;
}

File: book-bazaar/src/app/model/book.ts
import {Author} from './author';
import {Category} from './category';
import {Publisher} from './publisher';
import {Genre} from './genre';

export class Book {
  constructor(
    public id?: number,
    public title?: string,
    public author?: Author,
    public category?: Category,
    public description?: string,
    public isbn?: string,
    public publisher?: Publisher,
    public releaseDate?: string,
    public price?: number,
    public bookGenres?: Genre[],
    public imageUrl?: string,
    public averageRating?: number,
    public totalReviews?: number
  ) {
  }

  public static fromObject(book: Book) {
    return new Book(
      book.id,
      book.title,
      book.author,
      book.category,
      book.description,
      book.isbn,
      book.publisher,
      book.releaseDate,
      book.price,
      book.bookGenres,
      book.imageUrl,
      book.averageRating,
      book.totalReviews,
    );
  }
}
